src/frontend/Ast.ml

```diff
@@ -18,6 +18,8 @@ type 'e index =
 (** Front-end function kinds *)
 type fun_kind = StanLib | UserDefined [@@deriving compare, sexp, hash]
 
+let compare_fun_kind _ _ = 0
+
```

Comment:
This is a bit ugly. It's just a feature of our decision to add function kinds even to the unchecked AST. Really they should be part of the type that is only added with type checking.

Reply:
Is it possible to define this function at a scope closer to the test so that it's not always defined as this for all users of the AST? (Assuming some users might want to do actual comparisons of the function kinds, possibly).

Reply:
Good idea. I will try. Not sure if the ppx stuff works by picking up local definitions or if it only looks for global definitions. I'm afraid it may be the latter.

Reply:
Hmm. It's what I was afraid of. There is no way of moving this definition to local scope using ppx, as ppx only works in global scope.

I still think that we ultimately would want to move away from having the FnKinds as a part of the application. They should be part of the function type of the function being applied if anything. That way, they would also only be present in the typed ast and there would be no problem here. Currently, we are just setting them to fake values in the untyped ast and only to the correct value during type checking. That seems bad to me. At least, it should be an optional FnKind then if we are going this route, so we can set it to nothing in the untyped AST. Or we can just remove them in the untyped AST altogether.

What would your preference be, @seantalts ?

Reply:
Hmm, I could see it making sense to add the function kind only in the MIR. Do you remember why we didn't do that before? What would that look like?

Reply:
I guess it's added to the AST as well because the check that needs to happen to see the FnKind of an application is rather complicated and is performed anyway during type checking: it requires a lookup of the user defined functions that are in scope as they can sometimes shadow Stan Math functions.

Reply:
Gotcha. So it sounds like you'd like to keep it the way it is where we add the type during semantic checking? Would you then propose making it optional?

Reply:
OK. I've now gone with the most type safe route of using another two level pattern for the FunKinds. That allows me to remove the hacky compare function on FunKinds altogether. The route of optional FunKinds just seems not very type safe to me.

Reply:
I think what you have now is not two-level actually, it's equivalent to Optional, right? Am I missing something? I think it might be cleaner as `option` because then you don't have to thread the `'f` type through everywhere...

Reply:
Oh wait - I guess the difference is that you can add the presence of the fun_kind to the type signature in an easy way with this parameter such that you never see code matching on Some or None, right? I think that is better, though I think we might be getting to the point where we need to start using full names for all of our tons of type parameters 😂

Reply:
This is different as it's forcing through the type system that the untyped AST basically always have None for the fun_kind and that the typed one always has Some _ for it.
The number of type parameters is definitely getting a bit large. Not sure I know great names for them though.